parser_library/src/workspaces/program_configuration_storage.cpp

```diff
@@ -0,0 +1,239 @@
+}
+
+program_configuration_storage::get_pgm_result program_configuration_storage::get_program(
+    const utils::resource::resource_location& file_location_normalized) const
+{
+    if (const auto* details = get_program_properties(file_location_normalized))
+        return { std::get_if<program>(&details->pgm_details), details->affiliation };
+
+    return { nullptr, cfg_affiliation::none };
+}
+
+std::unordered_map<std::string, bool, utils::hashers::string_hasher, std::equal_to<>>
+program_configuration_storage::get_categorized_missing_pgroups(const utils::resource::resource_location& config_file_rl,
+    const std::vector<utils::resource::resource_location>& opened_files) const
+{
+    auto missing_proc_grps_it = m_missing_proc_grps.find(config_file_rl);
+    if (missing_proc_grps_it == m_missing_proc_grps.end())
+        return {};
+
+    std::unordered_map<std::string, bool, utils::hashers::string_hasher, std::equal_to<>> categorized_missing_pgroups;
+
+    for (const auto& missing_pgroup : missing_proc_grps_it->second)
+        categorized_missing_pgroups[missing_pgroup] = false;
+
+    for (const auto& opened_file : opened_files)
+    {
+        if (const auto missing_details = get_missing_pgroup_details(opened_file))
+            categorized_missing_pgroups[missing_details->pgroup_name] = true;
+    }
+
+    return categorized_missing_pgroups;
+}
+
+void program_configuration_storage::remove_conf(const void* tag)
+{
+    std::erase_if(m_exact_match, [&tag](const auto& e) { return e.second.tag == tag; });
+    std::erase_if(m_regex_pgm_conf, [&tag](const auto& e) { return e.first.tag == tag; });
+    std::erase_if(m_regex_b4g_json, [&tag](const auto& e) { return e.first.tag == tag; });
+}
+
+void program_configuration_storage::prune_external_processor_groups(const utils::resource::resource_location& location)
+{
+    static constexpr auto is_external = [](const program_properties& pgm_props) {
+        const auto* pgm = std::get_if<program>(&pgm_props.pgm_details);
+        return pgm && pgm->external;
+    };
+
+    if (!location.empty())
+    {
+        if (auto p = m_exact_match.find(location.lexically_normal());
+            p != m_exact_match.end() && is_external(p->second))
+            m_exact_match.erase(p);
+    }
+    else
+        std::erase_if(m_exact_match, [](const auto& p) { return is_external(p.second); });
+}
+
+void program_configuration_storage::clear()
+{
+    m_exact_match.clear();
+    m_regex_pgm_conf.clear();
+    m_regex_b4g_json.clear();
+    m_missing_proc_grps.clear();
+    m_cached_missing_proc_grps_it = {};
+}
+
+program_configuration_storage::missing_pgroup_details program_configuration_storage::new_missing_pgroup_helper(
+    std::string missing_pgroup_name, utils::resource::resource_location config_rl)
+{
+    if (!m_cached_missing_proc_grps_it || (*m_cached_missing_proc_grps_it)->first != config_rl)
+        m_cached_missing_proc_grps_it = m_missing_proc_grps.try_emplace(config_rl).first;
```

Comment:
The problem I'm trying to solve here is that the function `new_missing_pgroup_helper` can potentially be called many times with the same `config_rl`.

I'm trying to avoid having to lookup the relevant item in `m_missing_proc_grps` by caching it in `m_cached_missing_proc_grps_it`

Reply:
`(*m_cached_missing_proc_grps_it)->first != config_rl` will be similarly complex as trying to find the entry in the hash table.